packages/flutter/test/widgets/editable_text_test.dart

```diff
@@ -6555,6 +6556,304 @@ void main() {
+          '"oldText": "",'
+          ' "deltaText": "let there be text",'
+          ' "deltaType" : "TextEditingDeltaType.insertion",'
+          ' "deltaStart": 0,'
+          ' "deltaEnd": 0,'
+          ' "selectionBase": 17,'
+          ' "selectionExtent": 17,'
+          ' "selectionAffinity" : "TextAffinity.downstream" ,'
+          ' "selectionIsDirectional": false,'
+          ' "composingBase": -1,'
+          ' "composingExtent": -1}';
+
+      const String jsonDeletionDelta = '{'
+          '"oldText": "let there be text",'
+          ' "deltaText": "",'
+          ' "deltaStart": 12,'
+          ' "deltaEnd": 17,'
+          ' "selectionBase": 12,'
+          ' "selectionExtent": 12,'
+          ' "selectionAffinity" : "TextAffinity.downstream" ,'
+          ' "selectionIsDirectional": false,'
+          ' "composingBase": -1,'
+          ' "composingExtent": -1}';
+
+      const String jsonReplacementDelta = '{'
+          '"oldText": "let there be",'
+          ' "deltaText": "b light",'
+          ' "deltaStart": 10,'
+          ' "deltaEnd": 12,'
+          ' "selectionBase": 17,'
+          ' "selectionExtent": 17,'
+          ' "selectionAffinity" : "TextAffinity.downstream" ,'
+          ' "selectionIsDirectional": false,'
+          ' "composingBase": -1,'
+          ' "composingExtent": -1}';
+
+      const String jsonNonTextUpdateDelta = '{'
+          '"oldText": "let there b light",'
+          ' "deltaText": "",'
+          ' "deltaStart": -1,'
+          ' "deltaEnd": -1,'
+          ' "selectionBase": 17,'
+          ' "selectionExtent": 17,'
+          ' "selectionAffinity" : "TextAffinity.downstream",'
+          ' "selectionIsDirectional": false,'
+          ' "composingBase": -1,'
+          ' "composingExtent": -1}';
+
+      final TextEditingDelta insertionDelta = TextEditingDelta.fromJSON(jsonDecode(jsonInsertionDelta) as Map<String, dynamic>);
+      final TextEditingDelta deletionDelta = TextEditingDelta.fromJSON(jsonDecode(jsonDeletionDelta) as Map<String, dynamic>);
+      final TextEditingDelta replacementDelta = TextEditingDelta.fromJSON(jsonDecode(jsonReplacementDelta) as Map<String, dynamic>);
+      final TextEditingDelta nonTextUpdateDelta = TextEditingDelta.fromJSON(jsonDecode(jsonNonTextUpdateDelta) as Map<String, dynamic>);
+      expect(insertionDelta.deltaType, TextEditingDeltaType.insertion);
+      expect(deletionDelta.deltaType, TextEditingDeltaType.deletion);
+      expect(replacementDelta.deltaType, TextEditingDeltaType.replacement);
+      expect(nonTextUpdateDelta.deltaType, TextEditingDeltaType.nonTextUpdate);
+
+      state.updateEditingValueWithDeltas(<TextEditingDelta>[insertionDelta, deletionDelta, replacementDelta, nonTextUpdateDelta]);
+      await tester.pump();
+      expect(controller.text, 'let there b light');
+      expect(controller.selection, nonTextUpdateDelta.selection);
+      expect(state.currentTextEditingValue.composing, nonTextUpdateDelta.composing);
+    });
+  });
```

Comment:
I think you should add some unit tests to services/text_input_test.dart that test TextEditingDelta.fromJSON directly since there is a lot of logic in there.  Just make a bunch of calls to it with various kinds of json data and expect it to create the correct TextEditingDelta.

Reply:
That sounds like a good idea to me. I'll still leave these tests in `editable_test_text.dart` as they test the `updateEditingValueWithDelta` function.